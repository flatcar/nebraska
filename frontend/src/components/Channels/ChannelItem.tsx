import ScheduleIcon from '@mui/icons-material/Schedule';
import { Box, Grid, Tooltip, useTheme } from '@mui/material';
import ListItem from '@mui/material/ListItem';
import ListItemSecondaryAction from '@mui/material/ListItemSecondaryAction';
import ListItemText from '@mui/material/ListItemText';
import makeStyles from '@mui/styles/makeStyles';
import { useTranslation } from 'react-i18next';
import { Channel } from '../../api/apiDataTypes';
import { makeLocaleTime } from '../../i18n/dateTime';
import { applicationsStore } from '../../stores/Stores';
import { ARCHES, cleanSemverVersion } from '../../utils/helpers';
import MoreMenu from '../common/MoreMenu';
import ChannelAvatar from './ChannelAvatar';

const useStyles = makeStyles({
  root: {
    margin: '0px',
  },
});

export interface ChannelItemProps {
  channel: Channel;
  /** The default is to display the arch. */
  showArch?: boolean;
  /** The default is to not display the arch. */
  isAppView?: boolean;
  /** When an update to the channel happens. */
  onChannelUpdate?: (channelID: string) => void;
}

export default function ChannelItem(props: ChannelItemProps) {
  const theme = useTheme();
  const classes = useStyles();
  const { t } = useTranslation();
  const { channel, showArch = true, isAppView = false, onChannelUpdate = null, ...others } = props;
  const name = channel.name;
  const version = channel.package
    ? cleanSemverVersion(channel.package.version)
    : t('channels|no_package');

  function deleteChannel() {
    const confirmationText = t('channels|confirm_delete_channel');
    if (window.confirm(confirmationText)) {
      applicationsStore().deleteChannel(channel.application_id, channel.id);
    }
  }

  function updateChannel() {
    if (onChannelUpdate) {
      onChannelUpdate(channel.id);
    }
  }

  function getSecondaryText() {
    let text = '';

    if (version) {
      text = cleanSemverVersion(version);
    }

    if (showArch) {
      if (text !== '') {
        text += ' ';
      }

      text += `(${ARCHES[channel.arch]})`;
    }
    const date = channel.package ? new Date(channel.package.created_ts) : null;
    return (
      <Box display="flex" ml={1}>
        <Box>{text}</Box>
        {date && (
          <Box pl={2}>
            <Box display="flex">
              <Box>
                <Tooltip title={t('channels|release_date') || ''}>
                  <ScheduleIcon fontSize="small" />
                </Tooltip>
              </Box>
              <Box pl={1}>
                {makeLocaleTime(date, {
                  showTime: false,
                  dateFormat: { year: 'numeric', month: '2-digit', day: '2-digit' },
                })}
              </Box>
            </Box>
          </Box>
        )}
      </Box>
    );
  }
  return (
    <ListItem component="div" {...others}>
      <Grid container spacing={2}>
        <Grid item>
          {isAppView ? (
            <ChannelAvatar color={channel.color} size={theme.spacing(1)} />
          ) : (
            <ChannelAvatar color={channel.color}>{name[0]}</ChannelAvatar>
          )}
        </Grid>
        <Grid item>
          <ListItemText
            primary={
              <Box display="flex" alignItems="center">
                <Box pl={1} display="inline-block">
                  {name}
                </Box>
              </Box>
            }
            secondary={getSecondaryText()}
            className={classes.root}
            disableTypography
          />
        </Grid>
      </Grid>
      {onChannelUpdate && (
        <ListItemSecondaryAction>
          <MoreMenu
            options={[
              { label: t('frequent|edit'), action: updateChannel },
              { label: t('frequent|delete'), action: deleteChannel },
            ]}
          />
        </ListItemSecondaryAction>
      )}
    </ListItem>
  );
}
