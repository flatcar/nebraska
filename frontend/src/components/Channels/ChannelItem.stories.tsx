import { Meta, StoryFn } from '@storybook/react';
import ChannelItem, { ChannelItemProps } from './ChannelItem';

export default {
  title: 'channels/ChannelItem',
  argTypes: {
    onChannelUpdate: { action: 'onChannelUpdate' },
  },
} as Meta;

const Template: StoryFn<ChannelItemProps> = args => <ChannelItem {...args} />;

export const ShowArch = {
  render: Template,

  args: {
    channel: {
      id: '30b6ffa6-e6dc-4a01-bea6-9ce7f1a5bb34',
      name: 'edge',
      color: '#f4ab3b',
      created_ts: '2015-09-19T07:09:34.265754+02:00',
      application_id: 'e96281a6-d1af-4bde-9a0a-97b76e56dc57',
      package_id: 'X-A-Y',
      package: {
        id: 'PACK_ID',
        type: 4,
        version: '1.11.3',
        url: 'https://github.com/kinvolk',
        filename: '',
        description: '',
        size: '',
        hash: '',
        created_ts: '2019-07-18T20:10:39.163326+05:30',
        channels_blacklist: null,
        application_id: 'df1c8bbb-f525-49ee-8c94-3ca548b42059',
        flatcar_action: null,
        arch: 0,
        extra_files: [],
      },
      arch: 2,
    },
    showArch: true,
  },
};

export const NoPackage = {
  render: Template,

  args: {
    channel: {
      id: '30b6ffa6-e6dc-4a01-bea6-9ce7f1a5bb34',
      name: 'edge',
      color: '#f4ab3b',
      created_ts: '2015-09-19T07:09:34.265754+02:00',
      application_id: 'e96281a6-d1af-4bde-9a0a-97b76e56dc57',
      package_id: null,
      package: null,
      arch: 2,
    },
    showArch: false,
  },
};

export const NoArch = {
  render: Template,

  args: {
    ...ShowArch.args,
    showArch: false,
  },
};

export const AppView = {
  render: Template,

  args: {
    ...ShowArch.args,
    isAppView: true,
  },
};
