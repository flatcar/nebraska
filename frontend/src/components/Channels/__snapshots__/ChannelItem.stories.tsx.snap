// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`channels/ChannelItem AppView smoke-test 1`] = `
<li class="MuiListItem-container css-xw6fwh-MuiListItem-container">
  <div class="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-secondaryAction css-xo5bya-MuiListItem-root">
    <div class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-zjhuaj-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root">
        <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1lcewry-MuiAvatar-root">
        </div>
      </div>
      <div class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root">
        <div class="MuiListItemText-root MuiListItemText-multiline css-1fnxsti-MuiListItemText-root">
          <div class="MuiBox-root css-70qvj9">
            <div class="MuiBox-root css-1xsd608">
              edge
            </div>
          </div>
          <div class="MuiBox-root css-15aqakd">
            <div class="MuiBox-root css-0">
              1.11.3 (ARM64)
            </div>
            <div class="MuiBox-root css-n085mf">
              <div class="MuiBox-root css-k008qs">
                <div class="MuiBox-root css-0">
                  <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
                       focusable="false"
                       aria-hidden="true"
                       viewbox="0 0 24 24"
                       data-testid="ScheduleIcon"
                       aria-label="Release date"
                       data-mui-internal-clone-element="true"
                  >
                    <path d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8">
                    </path>
                    <path d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z">
                    </path>
                  </svg>
                </div>
                <div class="MuiBox-root css-1iuj5ih">
                  07/18/2019
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="MuiListItemSecondaryAction-root css-zezwi5-MuiListItemSecondaryAction-root">
    <div>
      <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeLarge css-geyil9-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
              aria-controls="simple-menu-4"
              aria-haspopup="true"
              aria-label="Open menu"
              data-testid="more-menu-open-button"
      >
        <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
             focusable="false"
             aria-hidden="true"
             viewbox="0 0 24 24"
             data-testid="MoreVertIcon"
        >
          <path d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2m0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2m0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2">
          </path>
        </svg>
      </button>
    </div>
  </div>
</li>
`;

exports[`channels/ChannelItem NoArch smoke-test 1`] = `
<li class="MuiListItem-container css-xw6fwh-MuiListItem-container">
  <div class="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-secondaryAction css-xo5bya-MuiListItem-root">
    <div class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-zjhuaj-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root">
        <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-10t1wcf-MuiAvatar-root">
          e
        </div>
      </div>
      <div class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root">
        <div class="MuiListItemText-root MuiListItemText-multiline css-1fnxsti-MuiListItemText-root">
          <div class="MuiBox-root css-70qvj9">
            <div class="MuiBox-root css-1xsd608">
              edge
            </div>
          </div>
          <div class="MuiBox-root css-15aqakd">
            <div class="MuiBox-root css-0">
              1.11.3
            </div>
            <div class="MuiBox-root css-n085mf">
              <div class="MuiBox-root css-k008qs">
                <div class="MuiBox-root css-0">
                  <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
                       focusable="false"
                       aria-hidden="true"
                       viewbox="0 0 24 24"
                       data-testid="ScheduleIcon"
                       aria-label="Release date"
                       data-mui-internal-clone-element="true"
                  >
                    <path d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8">
                    </path>
                    <path d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z">
                    </path>
                  </svg>
                </div>
                <div class="MuiBox-root css-1iuj5ih">
                  07/18/2019
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="MuiListItemSecondaryAction-root css-zezwi5-MuiListItemSecondaryAction-root">
    <div>
      <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeLarge css-geyil9-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
              aria-controls="simple-menu-3"
              aria-haspopup="true"
              aria-label="Open menu"
              data-testid="more-menu-open-button"
      >
        <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
             focusable="false"
             aria-hidden="true"
             viewbox="0 0 24 24"
             data-testid="MoreVertIcon"
        >
          <path d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2m0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2m0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2">
          </path>
        </svg>
      </button>
    </div>
  </div>
</li>
`;

exports[`channels/ChannelItem NoPackage smoke-test 1`] = `
<li class="MuiListItem-container css-xw6fwh-MuiListItem-container">
  <div class="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-secondaryAction css-xo5bya-MuiListItem-root">
    <div class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-zjhuaj-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root">
        <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-10t1wcf-MuiAvatar-root">
          e
        </div>
      </div>
      <div class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root">
        <div class="MuiListItemText-root MuiListItemText-multiline css-1fnxsti-MuiListItemText-root">
          <div class="MuiBox-root css-70qvj9">
            <div class="MuiBox-root css-1xsd608">
              edge
            </div>
          </div>
          <div class="MuiBox-root css-15aqakd">
            <div class="MuiBox-root css-0">
              No package
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="MuiListItemSecondaryAction-root css-zezwi5-MuiListItemSecondaryAction-root">
    <div>
      <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeLarge css-geyil9-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
              aria-controls="simple-menu-2"
              aria-haspopup="true"
              aria-label="Open menu"
              data-testid="more-menu-open-button"
      >
        <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
             focusable="false"
             aria-hidden="true"
             viewbox="0 0 24 24"
             data-testid="MoreVertIcon"
        >
          <path d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2m0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2m0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2">
          </path>
        </svg>
      </button>
    </div>
  </div>
</li>
`;

exports[`channels/ChannelItem ShowArch smoke-test 1`] = `
<li class="MuiListItem-container css-xw6fwh-MuiListItem-container">
  <div class="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-secondaryAction css-xo5bya-MuiListItem-root">
    <div class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-zjhuaj-MuiGrid-root">
      <div class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root">
        <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-10t1wcf-MuiAvatar-root">
          e
        </div>
      </div>
      <div class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root">
        <div class="MuiListItemText-root MuiListItemText-multiline css-1fnxsti-MuiListItemText-root">
          <div class="MuiBox-root css-70qvj9">
            <div class="MuiBox-root css-1xsd608">
              edge
            </div>
          </div>
          <div class="MuiBox-root css-15aqakd">
            <div class="MuiBox-root css-0">
              1.11.3 (ARM64)
            </div>
            <div class="MuiBox-root css-n085mf">
              <div class="MuiBox-root css-k008qs">
                <div class="MuiBox-root css-0">
                  <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
                       focusable="false"
                       aria-hidden="true"
                       viewbox="0 0 24 24"
                       data-testid="ScheduleIcon"
                       aria-label="Release date"
                       data-mui-internal-clone-element="true"
                  >
                    <path d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8">
                    </path>
                    <path d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z">
                    </path>
                  </svg>
                </div>
                <div class="MuiBox-root css-1iuj5ih">
                  07/18/2019
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="MuiListItemSecondaryAction-root css-zezwi5-MuiListItemSecondaryAction-root">
    <div>
      <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeLarge css-geyil9-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
              aria-controls="simple-menu-1"
              aria-haspopup="true"
              aria-label="Open menu"
              data-testid="more-menu-open-button"
      >
        <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
             focusable="false"
             aria-hidden="true"
             viewbox="0 0 24 24"
             data-testid="MoreVertIcon"
        >
          <path d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2m0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2m0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2">
          </path>
        </svg>
      </button>
    </div>
  </div>
</li>
`;
